### PR TITLE
Dynamic query and aggregation field

### PR DESCRIPTION
**Pull request by rafiqueud**
Dynamic query and aggregation field in search

### CODE REVIEW

File: lib/src/elastic_client_impl.dart [hunk 1]
```diff
@@ -102,15 +102,16 @@ class Client {
     return rs.bodyAsMap['deleted'] as int ?? 0;
   }
 
-  Future<SearchResult> search(String index, String type, Map query,
-      {int offset, int limit, bool fetchSource = false, Map suggest}) async {
+  Future<SearchResult> search(String index, String type, Map<String, dynamic> query,
+      {int offset, int limit, bool fetchSource = false, Map suggest, Map aggs, bool rawQuery = false}) async {
     final path = [index, type, '_search'];
-    final map = {
+    final map = rawQuery ? query: {
```
Comment (isoos) [inline lib/src/elastic_client_impl.dart]: why do you need the `rawQuery`?

Comment (rafiqueud) [reply lib/src/elastic_client_impl.dart]: In the latest versions of Elastic Database the query format change, and with your hardCoded {"query": query} will be common to be unable to use in that way.

I have passed by some problems with the query serialization and the deserialization of the return with differentes elastic server versions.
This situations has obrigate me to change your elastic_client code to be possible to use.

Another suggestion is to be possible to use a rawQuery and be possible to has the rawReturn.

Comment (isoos) [reply lib/src/elastic_client_impl.dart]: Hm, in that case, wouldn't it be better to make it as a client-level flag that you set at the constructor? I fear that passing it in each `query()` makes it cumbersome...

With that, how does logic change if you do have `rawQuery = true` and `offset` + `limit` specified? Could you also handle that case?

If you can copy here the query objects that were working and no longer working, I'd help with the code too.

Comment (rafiqueud) [reply lib/src/elastic_client_impl.dart]: Look my last commit, the total_count return in elastic v7.0 its a Map.
Think that in the future the query will have more than aggregations, limit, and other things, and the user want to build your own query.

Should be cool if the user can make your own map of query and send it as raw and dynamic map and get the response as a dynamic json. To treat like he want.

Comment (isoos) [reply lib/src/elastic_client_impl.dart]: I see it now, okay. Couple of things before I can merge it, first of all:

What do you think: would it make sense to have a separate `searchRaw` (or `rawSearch`) method? The current search would invoke it with the current query construction, while other could call it directly. That way the `bool rawQuery` would not be necessary.

The alternative is to also check if `rawQuery=true` and any other param is specified. E.g. `rawQuery=true` and `limit = 10` should throw an error to the client (and not swallow the lost parameter silently).

File: lib/src/elastic_client_impl.dart [hunk 2]
```diff
@@ -205,11 +193,11 @@ class ElasticDocHit {
 }
 
 abstract class Query {
-  static Map matchAll() => {'match_all': {}};
+  static Map<String, dynamic> matchAll() => {'match_all': {}};
```
Comment (isoos) [inline lib/src/elastic_client_impl.dart:196]: I believe in these cases the return value should be typed:

```
static Map<String, dynamic> matchAll() => <String, dynamic>{'match_all': {}};

// maybe just this, as type will be inferred
static matchAll() => <String, dynamic>{'match_all': {}};
```

File: lib/src/elastic_client_impl.dart [hunk 3]
```diff
@@ -205,11 +193,11 @@ class ElasticDocHit {
...
 
-  static Map matchNone() => {'match_none': {}};
+  static Map<String, dynamic> matchNone() => {'match_none': {}};
```
Comment (isoos) [inline lib/src/elastic_client_impl.dart:198]: same typed return value here and below...

**Comment by isoos** [comment]
@rafiqueud the PR seems solid, my only question is the `searchRaw` vs. not swallowing silent errors, and the typed return maps. Please fix those for better consistency, thank you!

File: lib/src/elastic_client_impl.dart (hunk 1 above)
Comment (olachris) [reply lib/src/elastic_client_impl.dart]: +1 for having this as a separate rawSearch method

**Comment by isoos** [comment]
@rafiqueud: Are you still interested completing the PR? No worries if this became cumbersome, I'll try my best to adopt in a new PR what we've discussed here.

File: lib/src/elastic_client_impl.dart (hunk 1 above)
Comment (rafiqueud) [reply lib/src/elastic_client_impl.dart]: +1 for having this as a separate rawSearch method

Comment (elgansayer) [reply lib/src/elastic_client_impl.dart]: nice. I just modified everything and added a load more queries and removed a lot of the doc stuff for the new versions.

https://github.com/elgansayer/elastic_client

Tbh i should probably clean it up and submit a PR.

Comment (isoos) [reply lib/src/elastic_client_impl.dart]: @elgansayer: I'd encourage you to make that PR!

**Pull request closed without merging**